lib/cosmos/api/cmd_api.rb

```diff
@@ -404,9 +417,48 @@ def cmd_implementation(range_check, hazardous_check, raw, method_name, *args, sc
         raise "ERROR: Invalid number of arguments (#{args.length}) passed to #{method_name}()"
       end
       authorize(permission: 'cmd', target_name: target_name, packet_name: cmd_name, scope: scope, token: token)
-      Store.instance.cmd_target(target_name, cmd_name, cmd_params, range_check, hazardous_check, raw, scope: scope)
-      [target_name, cmd_name, cmd_params]
+      TargetModel.packet(target_name, cmd_name, type: :CMD, scope: scope)
+
+      timeout_ms = 5000  # TODO: This 5 second timeout is arbitrary ... what should it be
+      ack_topic = "#{scope}__ACKCMDTARGET__#{target_name}"
+      Store.update_topic_offsets([ack_topic])
+      cmd_id = Store.write_topic("#{scope}__CMDTARGET__#{target_name}", {
+        'target_name' => target_name,
+        'cmd_name' => cmd_name,
+        'cmd_params' => JSON.generate(cmd_params.as_json),
+        'range_check' => range_check,
+        'hazardous_check' => hazardous_check,
+        'raw' => raw })
+      (timeout_ms / 100).times do
+        Topic.read_topics([ack_topic]) do |topic, msg_id, msg_hash, redis|
+          # Logger.debug("topic:#{topic} id:#{msg_id} hash:#{msg_hash.inspect}")
+          if msg_hash["id"] == cmd_id
+            # Logger.debug "Ack Received topic:#{topic} id:#{msg_id} msg:#{msg_hash.inspect}"
+            if msg_hash["result"] == "SUCCESS"
+              return [target_name, cmd_name, cmd_params]
+            # Check for HazardousError which is a special case
+            elsif msg_hash["result"].include?("HazardousError")
+              raise_hazardous_error(msg_hash, target_name, cmd_name, cmd_params)
+            else
+              raise msg_hash["result"]
+            end
+          end
+        end
+      end
```

Comment:
This code feels like it should probably be in one of the topic classes?

Reply:
Agreed. I wasn't as familiar with the topics at the time and just wanted to move it out of Store. Will do in next PR.